tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc

```diff
@@ -909,38 +908,52 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
       return false;
     }
 
-    // Sequentially apply the collected unary, dynamic-slice and pad ops to the
-    // unconverted and unscaled operands.
-    auto shift_unary_ops =
+    // Sequentially apply the collected unary, dynamic-slice, pad and select ops
+    // to the unconverted and unscaled operands.
+    auto shift_ops =
         [&instr](HloInstruction *&x,
-                 std::vector<HloInstruction *> &x_unary_ops) -> void {
-      for (HloInstruction *unary_op : x_unary_ops) {
+                 std::vector<std::pair<HloInstruction *, int>> &x_ops) -> void {
+      for (std::pair<HloInstruction *, int> op : x_ops) {
         std::vector<HloInstruction *> operands = {x};
         // Insert the additional operands of dynamic-slice ops.
-        if (unary_op->opcode() == HloOpcode::kDynamicSlice) {
-          for (int i = 1; i < unary_op->operand_count(); ++i) {
-            operands.emplace_back(unary_op->mutable_operand(i));
+        if (op.first->opcode() == HloOpcode::kDynamicSlice) {
+          for (int i = 1; i < op.first->operand_count(); ++i) {
+            operands.emplace_back(op.first->mutable_operand(i));
           }
         }
         // Convert the second operand of pad ops.
-        if (unary_op->opcode() == HloOpcode::kPad) {
+        if (op.first->opcode() == HloOpcode::kPad) {
           HloInstruction *convert =
               instr->AddInstruction(HloInstruction::CreateConvert(
-                  ShapeUtil::ChangeElementType(unary_op->operand(1)->shape(),
+                  ShapeUtil::ChangeElementType(op.first->operand(1)->shape(),
                                                x->shape().element_type()),
-                  unary_op->mutable_operand(1)));
+                  op.first->mutable_operand(1)));
           operands.emplace_back(convert);
         }
-        x = instr->AddInstruction(unary_op->CloneWithNewOperands(
+        // Convert and insert the additional operands of select ops.
+        if (op.first->opcode() == HloOpcode::kSelect) {
+          // The first operand is the predicate.
+          operands.emplace(operands.begin(), op.first->mutable_operand(0));
+          // Convert the remaining operand.
+          int operand_idx = op.second == 2 ? 1 : 2;
+          HloInstruction *convert =
+              instr->AddInstruction(HloInstruction::CreateConvert(
+                  ShapeUtil::ChangeElementType(
+                      op.first->operand(operand_idx)->shape(),
+                      x->shape().element_type()),
+                  op.first->mutable_operand(operand_idx)));
```

Comment:
You are changing

```
x->convert->multiply->select->dot
```

into

```
x->select->convert->multiply->dot
```

But now if the other argument to select is selected (the argument that's not `x` and not the predicate), it will be multiplied while before, it wasn't multiplied, thus changing correctness. Inserting a divide to undo the multiply doesn't work great either, since the divide will be operating on an FP8 tensor and may cause overflow/underflow.

I don't see any way of handling this case in a general way. The test case you have uses a constant of zero, in which case there is no overflow/underflow risk, nor does the multiplication have an effect. If in the real FP8 model you're looking at, the select has a zero, I would only add Select to the subgraph in `IsSupportedF8Pattern` if the other argument is zero. Otherwise, we'll have to discuss further how to support this.